syntax = "proto3";

option java_package = "org.retrostore.client.common.proto";
option java_outer_classname = "ApiProtos";
option java_multiple_files = true;

// Response for APIs that receive a list of apps.
message ApiResponseApps {
    // Whether the request was a success.
    bool success = 1;
    // An optional (error) message, human-readable.
    string message = 2;
    // A list of apps returned by the request.
    repeated App app = 3;
}

message ApiResponseMediaImages {
    // Whether the request was a success.
    bool success = 1;
    // An optional (error) message, human-readable.
    string message = 2;
    // The media images returnes by this respone.
    repeated MediaImage mediaImage = 3;
}

// A RetroStore app item.
message App {
    // The ID to uniquely identify an app.
    string id = 1;
    // The name of the app.
    string name = 2;
    // The human readable version of this app.
    string version = 3;
    // The description of this app.
    string description = 4;
    // The original release year of the app.
    int32 release_year = 5;
    // URLs to screenshots for this app.
    repeated string screenshot_url = 6;
    // The author of the app (not the uploader).
    string author = 7;
    // Extension set for TRS80 apps.
    Trs80Extension ext_trs80 = 8;

    // NEXT ID: 9;
}

// Parameters specific to TRS-80 apps.
message Trs80Extension {
    // The TRS-80 model type.
    Trs80Model model = 1;

    // The TRS-80 model types.
    enum Trs80Model {
        UNKNOWN_MODEL = 0;
        MODEL_I = 1;
        MODEL_III = 2;
        MODEL_4 = 3;
        MODEL_4P = 4;
    }
}

// A media image for an app.
message MediaImage {
    // The type of this media image.
    MediaType type = 1;
    // The file name of this media image.
    string filename = 2;
    // The actual data of this media image.
    bytes data = 3;
    // When the image was uploaded.
    int64 uploadTime = 4;
    // An optional description of this media image describing its contents.
    string description = 5;
}

// Types of MediaImages.
enum MediaType {
    UNKNOWN = 0;
    DISK = 1;
    CASSETTE = 2;
    COMMAND = 3;
    BASIC = 4;
}
